### PR TITLE
Ability to register new blockchain and wallet

### PR DESCRIPTION
**Pull request by ritesh-skyach**
Ability to list blockchains and wallets from admin panel #1412
Ability to register a new blockchain / wallet from admin panel #1413

### CODE REVIEW

File: app/controllers/admin/blockchains_controller.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      @blockchains = Blockchain.all
+    end
+
+    def new
+      @blockchain = Blockchain.new
+      render :show
+    end
+
+    def create
+      @blockchain = Blockchain.new
```
Comment (tamadamas) [inline app/controllers/admin/blockchains_controller.rb]: Why do you use assign_attributes here istead of `@blockchain = Blockchain.create(blockchain_params)`

File: app/controllers/admin/blockchains_controller.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      @blockchain.assign_attributes(blockchain_params)
+      if @blockchain.save
+        redirect_to admin_blockchains_path
+      else
+        flash[:alert] = @blockchain.errors.full_messages.first
+        render :show
+      end
+    end
+
+    def show
```
Comment (tamadamas) [inline app/controllers/admin/blockchains_controller.rb]: `show` action should be after `index` action before `new`

File: app/controllers/admin/blockchains_controller.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      @blockchain = Blockchain.find(params[:id])
+    end
+
+    def update
+      @blockchain = Blockchain.find(params[:id])
+      if @blockchain.update(blockchain_params)
+        redirect_to admin_blockchains_path
+      else
+        flash[:alert] = @blockchain.errors.full_messages.first
```
Comment (tamadamas) [inline app/controllers/admin/blockchains_controller.rb]: Just first error? What if model has two errors?

File: app/controllers/admin/blockchains_controller.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          server
+          height
+          deposit_confirmations
+          withdraw_confirmations
+          explorer_address
+          explorer_transaction
+          status
+          case_sensitive
+      ]
+
```
Comment (tamadamas) [inline app/controllers/admin/blockchains_controller.rb]: remove blank lines

File: app/controllers/admin/blockchains_controller.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+module Admin
+  class BlockchainsController < BaseController
+    def index
+      @blockchains = Blockchain.all
```
Comment (tamadamas) [inline app/controllers/admin/blockchains_controller.rb]: Should we add a pagination here?

File: app/controllers/admin/wallets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    def new
+      @wallet = Wallet.new
+      render :show
+    end
+
+    def create
+      @wallet = Wallet.new
+      @wallet.assign_attributes(wallet_params)
```
Comment (tamadamas) [inline app/controllers/admin/wallets_controller.rb]: Why do you use assign_attributes here istead of @wallet = Blockchain.create(wallet_params)

File: app/controllers/admin/wallets_controller.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      if @wallet.save
+        redirect_to admin_wallets_path
+      else
+        flash[:alert] = @wallet.errors.full_messages.first
```
Comment (tamadamas) [inline app/controllers/admin/wallets_controller.rb]: same as above

File: app/controllers/admin/wallets_controller.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    def show
+      @wallet = Wallet.find(params[:id])
+    end
+
+    def update
+      @wallet = Wallet.find(params[:id])
+      if @wallet.update(wallet_params)
+        redirect_to admin_wallets_path
+      else
+        flash[:alert] = @wallet.errors.full_messages.first
```
Comment (tamadamas) [inline app/controllers/admin/wallets_controller.rb]: same as above

File: app/controllers/admin/wallets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+      %i[
+          currency_id
+          name
+          address
+          kind
+          nsig
+          parent
+          status
+      ]
+
```
Comment (tamadamas) [inline app/controllers/admin/wallets_controller.rb]: blank lines

File: app/views/admin/blockchains/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,46 @@
+<div class="row">
+  <div class="col-6">
+    <h3>Blockchains</h3>
+  </div>
+  <div class="col-6">
+    <%= link_to 'New currency', new_admin_blockchain_path, class: "btn btn-primary pull-right mb-3" %>
```
Comment (tamadamas) [inline app/views/admin/blockchains/index.html.erb]: should we check CanCan rule here?

File: app/views/admin/wallets/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,48 @@
+<div class="row">
+  <div class="col-6">
+    <h3>Wallets</h3>
+  </div>
+  <div class="col-6">
+    <%= link_to 'New currency', new_admin_wallet_path, class: "btn btn-primary pull-right mb-3" %>
```
Comment (tamadamas) [inline app/views/admin/wallets/index.html.erb]: should we check CanCan rule here?

File: spec/controllers/admin/blockchains_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+  before { session[:member_id] = member.id }
+
+  describe '#create' do
+    it 'creates blockchain with valid attributes' do
+      expect do
+        post :create, blockchain: attributes
+        expect(response).to redirect_to admin_blockchains_path
+      end.to change(Blockchain, :count)
+      blockchain = Blockchain.find_by_key('eth-rinkeby-new')
```
Comment (tamadamas) [inline spec/controllers/admin/blockchains_controller_spec.rb]: 'eth-rinkeby-new' should be constant.
You can get all attributes and use one `expect` to match them all

File: spec/controllers/admin/blockchains_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        withdraw_confirmations:           3,
+        explorer_address:                 'https://www.blocktrail.com/BCC/address/\#{address}',
+        explorer_transaction:             'https://blockchain.info/tx/\#{txid}',
+        status:                           'active',
+        case_sensitive:                   true }
+    end
+
+    before { request.env['HTTP_REFERER'] = '/admin/blockchains' }
+
+    it 'updates blockchain attributes' do
```
Comment (tamadamas) [inline spec/controllers/admin/blockchains_controller_spec.rb:50]: what do you test here? You tested `create` logic above
It should test "update" logic.

You can get all attributes and use one `expect` to match them all

File: spec/controllers/admin/blockchains_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    end
+  end
+
+  describe '#destroy' do
+    it 'doesn\'t support deletion of blockchain' do
+      expect { delete :destroy, id: existing_blockchain.id }.to raise_error(ActionController::UrlGenerationError)
+    end
+  end
+
+  describe 'routes' do
```
Comment (tamadamas) [inline spec/controllers/admin/blockchains_controller_spec.rb]: why routes test here? It should be another test file with `type: :routing`

File: spec/controllers/admin/wallets_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        address:            '249048804499541338815845805798634312140346616732',
+        kind:               'warm',
+        nsig:               5,
+        status:             'disabled' }
+    end
+
+    before { request.env['HTTP_REFERER'] = '/admin/wallets' }
+
+    it 'updates wallet attributes' do
+      post :create, wallet: attributes
```
Comment (tamadamas) [inline spec/controllers/admin/wallets_controller_spec.rb:44]: same as above

File: spec/controllers/admin/wallets_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+  describe '#destroy' do
+    it 'doesn\'t support deletion of wallet' do
+      expect { delete :destroy, id: existing_wallet.id }.to raise_error(ActionController::UrlGenerationError)
+    end
+  end
+
+  describe 'routes' do
+    let(:base_route) { '/admin/wallets' }
+    it 'routes to WalletsController' do
```
Comment (tamadamas) [inline spec/controllers/admin/wallets_controller_spec.rb]: same as above

**Review by tamadamas — CHANGES_REQUESTED** [review]

File: spec/controllers/admin/blockchains_controller_spec.rb (hunk 12 above)
Comment (ritesh-skyach) [reply spec/controllers/admin/blockchains_controller_spec.rb]: @rxx I didn't get what do you mean by `'eth-rinkeby-new' should be constant.` ?

Comment (tamadamas) [reply spec/controllers/admin/blockchains_controller_spec.rb]: `let(:blockchain_key) { 'eth-rinkeby-new' } `

Comment (tamadamas) [reply spec/controllers/admin/blockchains_controller_spec.rb]: @ritesh-skyach

**Comment by ysv** [comment]
Merged without rebase

**Pull request merged**